RFC: Implement a modal goto panel

pfitzseb opened the pull request:
Pretty much [this](https://github.com/JunoLab/atom-julia-client/pull/82), except more general.

I'm not quire sure how to handle the case when only one item exists -- we can't know that before the promise is fulfilled, and that should arguably only happen _after_ the goto-panel is displayed. It'd be a bit awkward to close it right after and jump straight to the def.

ToDo:
- [ ] Specs!

pfitzseb committed:
Implement a modal goto panel

pfitzseb committed:
cs syntax is a bit confusing sometimes :)

pfitzseb committed:
styling

MikeInnes commented:
Looking good! RE the single-item issue, how about if the panel isn't shown at all before the promise is fulfilled? The loading indicator will show while it's waiting, but tbh it should barely matter since the request should basically always be instant.

MikeInnes commented on lib/gotodef.coffee:
```diff
@@ -0,0 +1,88 @@
+  promise.then ((items) -> @view.setItems items),
+                (error) -> @view.setError error
+
+class goToView extends SelectListView
```
This class name style sticks out a bit, doesn't it?

pfitzseb replied:
True. Would `GotoView` look better?

MikeInnes commented on lib/gotodef.coffee:
```diff
@@ -0,0 +1,88 @@
+  @view.setLoading "Loading..."
+  @view.show()
+  promise.then ((items) -> @view.setItems items),
+                (error) -> @view.setError error
```
This might be clearer as

``` coffeescript
promise
  .then ...
  .catch ...
```

MikeInnes commented on lib/gotodef.coffee:
```diff
@@ -0,0 +1,88 @@
+
+module.exports =
+goto: (promise) ->
+  @view ?= new MethodView()
```
This doesn't seem to be defined anywhere?

pfitzseb commented:
Yeah, sounds good. Where should the loading indicator be activated -- should that happen on a call to `goto` or should it be the responsibility of the caller? I've gone with the latter option for now, but should be easy to change.

Also, thanks for the review!

pfitzseb committed:
bit of refactoring

pfitzseb committed:
keeps docs up to date

pfitzseb commented:
There should also be some unobstrusive mechanism for displaying errors and something like `No definitions found.` on `items.length == 0`. Do we already have something like that? I'd rather not use the default Atom notifications, they're too much of an eyecatcher...

MikeInnes commented:
I tend to agree, but I don't think there's a realistic alternative right now. In future we might want to have some kind of status bar notification, but that's already getting pretty full. Something toast-like might be nice.

I think `no definitions found == no response` would be OK for now, so long as it's completely predictable.

pfitzseb committed:
update error mode
yes, that's much less elegant than before

pfitzseb committed:
update this again

pfitzseb committed:
whitespace

pfitzseb commented:
Alright, this still needs specs, but should otherwise be good to go, I think.

pfitzseb committed:
whitespace

pfitzseb commented:
I've rebased and updated the PR (and [this](https://github.com/JunoLab/atom-julia-client/pull/149) and [that](https://github.com/JunoLab/Atom.jl/pull/20) as well).
Whaddaya think of merging, @MikeInnes?

Yes, I realize the lack of tests is a bit hypocritical... :)

MikeInnes commented:
> Yes, I realize the lack of tests is a bit hypocritical... :)

Haha no worries. I think what makes sense is to do a big push for tests once things are a bit more stable.

This looks good! Merge at your leisure :)

pfitzseb committed:
allow both promises and symbolTables as input
also light refactoring

pfitzseb committed:
rename stuff
